### ISSUE TITLE
number column: variable width

### ISSUE THREAD

**Issue by m5w** [neovim/neovim#4722]
This is a suggestion.
- Neovim version: 0.1.4 (4dcd19d9bc2417051ddbda177010ca8c0cb2cf73)
-  Vim behaves differently? Vim version: N/a
- Operating system/version: Debian "jessie"
- Terminal name/version: GNOME Terminal 3.14.1
- `$TERM`: xterm
### Actual behaviour

The maximum number of columns the user can declare to be used for the line number is 9 (since the maximum value of 'numberwidth' is 10, and 'numberwidth' is one column greater than those used for the line number). However, the line numbers can widen to ten columns if a file with over 999,999,999 lines is read.
### Expected behaviour

Perhaps the user should be able to use more than 9 columns for the line number, since this is done anyway if necessary. As 'numberwidth' is stored in a long---see [this](https://github.com/neovim/neovim/blob/b02ba11cb156e094d3f6f12fc7c79a983b6df68d/src/nvim/buffer_defs.h#L198)---a reasonable value might be just one column greater, as the standard only requires a long to have a maximum value of 2 ^ 31 - 1 , or 2,147,483,647. (Of course, a long on a 64-bit machine could be as large as 9,223,372,036,854,775,807, but I know of no machine with 9 exabytes of memory---though there are machines with 10GB and more. Nonetheless, I've tested changing the maximum value of 'numberwidth' hard-coded around [here](https://github.com/neovim/neovim/blob/d542de4a76dd9e600ebcf1405efdc9d3090ad9a8/src/nvim/option.c#L4121) and have noticed no problems, so perhaps the maximum could be changed to 20 columns for good measure.)
### Steps to reproduce using `nvim -u NORC`

Edit any file with more than 999,999,999 lines. You could do the following, for example:

```
# Generate a file with 2 ^ n newlines.
n=31
cat >1 <<EOF

EOF
for i in {1.."${n}"}; do cat 1 1 >2 && mv 2 1; done # courtesy of <http://www.skorks.com/2010/03/how-to-quickly-generate-a-large-file-on-the-command-line-with-linux/>
nvim -u NORC 1
:se number
:se numberwidth=10
```

**Comment by ZyX-I** [comment]
I would suggest changing that value to NUMBUFLEN: this constant is used around the code to allocate buffers to print numbers to them (including printing 64-bit unsigned integers in base 2) and thus is guaranteed to exceed anything user may want to set &numberwidth to.

Though this cannot be documented without creating one of the following situations:

1. WTF are you talking about? (user reading documentation if user documentation simply references NUMBUFLEN)
2. Blah, forgot to update the doc. (developer wanting to e.g. support 128-bit integers, assuming doxygen documentation contains “when changing this constant please update `:h 'numberwidth'`”)
3. O! I can set &numberwidth to 129 while documentation says 65, is this a bug? (same as 2, but without note in doxygen documentation or with both reviewer and author missing it)

**Comment by justinmk** [comment]
See also https://github.com/neovim/neovim/issues/990

**Comment by shtanton** [comment]
PR #10314 just updated the extra string to be 57 bytes, so how about 56 as the max?
Or I could increase the size of extra again to NUMBUFLEN, and use NUMBUFLEN-1 for the max. This connects them so increasing NUMBUFLEN in future will increase both by the right amount.

**Comment by shtanton** [comment]
I made a PR on [vim](https://github.com/vim/vim/pull/4584) for this

**Comment by shtanton** [comment]
This is now in vim 8.1.1614
What is the procedure for copying this into neovim?

**Comment by justinmk** [comment]
@shtanton procedure is here: https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim

**Comment by Billy4195** [comment]
@justinmk @mhinz I think this issue can be closed because the vim patch 8.1.1614 has been merged at 8062e6ff88247e434890dfdfcc25ae293dafc257
